File: packages/quant-ts/src/minvar.ts
/**
 * Global minimum-variance portfolio — TypeScript mirror of
 * `backend/quant/minvar.py`.
 *
 * - Unconstrained: `w = Σ⁻¹𝟏 / (𝟏ᵀΣ⁻¹𝟏)`.
 * - Long-only: active-set iteration (same structure as tangency).
 */

import { OptimizerInfeasibleError } from "./errors.js";
import { dot, ensurePsdCovariance, matVec, solve, sumVec } from "./linalg.js";
import type { ORP } from "./types.js";

export const SUM_TOL = 1e-9;

function mvpUnconstrained(cov: number[][]): number[] {
  const n = cov.length;
  const ones = new Array<number>(n).fill(1);
  const z = solve(cov, ones);
  const s = sumVec(z);
  if (!Number.isFinite(s) || Math.abs(s) < 1e-14) {
    throw new OptimizerInfeasibleError(
      "unconstrained MVP normalization is degenerate",
      { sum: s },
    );
  }
  return z.map((v) => v / s);
}

function subMatrix(matrix: readonly (readonly number[])[], idx: readonly number[]): number[][] {
  return idx.map((i) => idx.map((j) => ((matrix[i] as readonly number[])[j] as number)));
}

function mvpLongOnly(cov: number[][]): number[] {
  const n = cov.length;
  let active = Array.from({ length: n }, (_, i) => i);

  for (let iter = 0; iter < n + 1; iter += 1) {
    if (active.length === 0) {
      throw new OptimizerInfeasibleError("long-only MVP active set became empty", {});
    }
    const subCov = subMatrix(cov, active);
    const subW = mvpUnconstrained(subCov);

    const minIdx = subW.reduce((iMin, v, i) => (v < (subW[iMin] as number) ? i : iMin), 0);
    const minVal = subW[minIdx] as number;
    if (minVal >= -1e-12) {
      const full = new Array<number>(n).fill(0);
      for (let k = 0; k < active.length; k += 1) {
        full[active[k] as number] = Math.max(subW[k] as number, 0);
      }
      const s = sumVec(full);
      if (s <= SUM_TOL) {
        throw new OptimizerInfeasibleError("long-only MVP produced zero weights", { sum: s });
      }
      return full.map((v) => v / s);
    }
    active = active.filter((_, k) => k !== minIdx);
  }
  throw new OptimizerInfeasibleError(
    "long-only MVP failed to converge within n iterations",
    { n },
  );
}

export interface MinimumVarianceInput {
  tickers: readonly string[];
  expectedReturns: readonly number[];
  covariance: number[][];
  riskFreeRate: number;
  allowShort: boolean;
  warnings?: string[];
}

export function minimumVariancePortfolio(input: MinimumVarianceInput): ORP {
  const { tickers, expectedReturns, covariance, riskFreeRate, allowShort, warnings } = input;
  if (expectedReturns.length !== tickers.length) {
    throw new Error(
      `tickers length ${tickers.length} != expectedReturns length ${expectedReturns.length}`,
    );
  }
  if (covariance.length !== tickers.length) {
    throw new Error(`covariance must be ${tickers.length}x${tickers.length}`);
  }
  const cov = ensurePsdCovariance(covariance, warnings);

  let w = allowShort ? mvpUnconstrained(cov) : mvpLongOnly(cov);

  const s = sumVec(w);
  if (Math.abs(s - 1) > SUM_TOL) w = w.map((x) => x / s);

  const expectedReturn = dot(expectedReturns, w);
  const variance = dot(w, matVec(cov, w));
  if (variance <= 0) {
    throw new OptimizerInfeasibleError("MVP variance came out non-positive", {
      variance,
    });
  }
  const stdDev = Math.sqrt(variance);
  const sharpe = (expectedReturn - riskFreeRate) / stdDev;

  const weights: Record<string, number> = {};
  for (let i = 0; i < tickers.length; i += 1) {
    weights[tickers[i] as string] = w[i] as number;
  }
  return { weights, expectedReturn, stdDev, variance, sharpe };
}

File: backend/quant/minvar.py
"""Global minimum-variance portfolio (MVP).

- **Unconstrained (``allow_short=True``).** Closed form
  ``w = Σ⁻¹𝟏 / (𝟏ᵀΣ⁻¹𝟏)``.
- **Long-only (``allow_short=False``).** ``min wᵀΣw`` subject to
  ``Σwᵢ = 1``, ``w ≥ 0``; solved with ``cvxpy``.

Returned as an :class:`quant.types.ORP` for symmetry with
:func:`quant.markowitz.optimize_markowitz` — the Sharpe field is a valid
Sharpe at this particular ``(E(r), σ, rᶠ)``, but is usually smaller than the
ORP's.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from .errors import OptimizerInfeasibleError
from .linalg import ensure_psd_covariance
from .types import ORP

SUM_TOL: float = 1e-9


def _mvp_unconstrained(covariance: NDArray[np.float64]) -> NDArray[np.float64]:
    n = covariance.shape[0]
    ones = np.ones(n, dtype=np.float64)
    z = np.linalg.solve(covariance, ones)
    s = float(np.sum(z))
    if not np.isfinite(s) or abs(s) < 1e-14:
        raise OptimizerInfeasibleError(
            "unconstrained MVP normalization is degenerate",
            {"sum": s},
        )
    return np.asarray(z / s, dtype=np.float64)


def _mvp_long_only(covariance: NDArray[np.float64]) -> NDArray[np.float64]:
    import cvxpy as cp

    n = covariance.shape[0]
    w = cp.Variable(n, nonneg=True)
    problem = cp.Problem(
        cp.Minimize(cp.quad_form(w, cp.psd_wrap(covariance))),
        [cp.sum(w) == 1.0],
    )
    try:
        problem.solve()
    except cp.error.SolverError as exc:  # pragma: no cover - solver-specific
        raise OptimizerInfeasibleError(
            "cvxpy solver raised during long-only MVP",
            {"error": str(exc)},
        ) from exc

    if problem.status not in {"optimal", "optimal_inaccurate"} or w.value is None:
        raise OptimizerInfeasibleError(
            "long-only MVP is infeasible or unbounded",
            {"status": str(problem.status)},
        )

    raw = np.asarray(w.value, dtype=np.float64).reshape(-1)
    raw = np.clip(raw, 0.0, None)
    s = float(raw.sum())
    if s <= SUM_TOL:
        raise OptimizerInfeasibleError(
            "long-only MVP produced a zero-weight vector",
            {"sum": s},
        )
    return raw / s


def minimum_variance_portfolio(
    tickers: list[str],
    expected_returns: NDArray[np.float64] | list[float],
    covariance: NDArray[np.float64] | list[list[float]],
    risk_free_rate: float,
    allow_short: bool,
    warnings: list[str] | None = None,
) -> ORP:
    """Solve for the global minimum-variance portfolio."""
    mu = np.asarray(expected_returns, dtype=np.float64).reshape(-1)
    cov_in = np.asarray(covariance, dtype=np.float64)
    if mu.shape[0] != len(tickers):
        raise ValueError(f"tickers length {len(tickers)} != expected_returns length {mu.shape[0]}")
    if cov_in.shape != (mu.shape[0], mu.shape[0]):
        raise ValueError(f"covariance must be ({mu.shape[0]},{mu.shape[0]}); got {cov_in.shape}")

    cov = ensure_psd_covariance(cov_in, warnings=warnings)

    weights_vec = _mvp_unconstrained(cov) if allow_short else _mvp_long_only(cov)

    w_sum = float(weights_vec.sum())
    if abs(w_sum - 1.0) > SUM_TOL:
        weights_vec = weights_vec / w_sum

    e_r = float(mu @ weights_vec)
    var = float(weights_vec @ cov @ weights_vec)
    if var <= 0.0:
        raise OptimizerInfeasibleError("MVP variance came out non-positive", {"variance": var})
    sd = float(np.sqrt(var))
    sharpe = (e_r - float(risk_free_rate)) / sd

    weights = {tickers[i]: float(weights_vec[i]) for i in range(len(tickers))}

    return ORP(
        weights=weights,
        expected_return=e_r,
        std_dev=sd,
        variance=var,
        sharpe=sharpe,
    )


__all__ = ["SUM_TOL", "minimum_variance_portfolio"]
